File: breakpilot-vscode/src/debugger/PauseScopedHandleRegistry.ts
import { randomUUID } from "node:crypto";

export type HandleDescriptor = {
  sessionId: string;
  pauseEpoch: number;
  dapVariablesReference: number;
  parentVariablesReference?: number;
  name: string;
  evaluateName?: string;
  threadId?: number;
  frameId?: number;
  modifiable?: boolean;
};

export class PauseScopedHandleRegistry {
  private readonly entries = new Map<string, HandleDescriptor>();
  private readonly maxEntries: number;

  constructor(maxEntries = 2048) {
    this.maxEntries = maxEntries;
  }

  register(descriptor: HandleDescriptor): string {
    const handle = `bpref_${randomUUID()}`;
    this.entries.set(handle, { ...descriptor });
    while (this.entries.size > Math.max(1, this.maxEntries)) {
      const oldest = this.entries.keys().next().value as string | undefined;
      if (!oldest) break;
      this.entries.delete(oldest);
    }
    return handle;
  }

  resolve(handle: string, sessionId: string, pauseEpoch: number): HandleDescriptor | undefined {
    const descriptor = this.entries.get(handle);
    if (!descriptor || descriptor.sessionId !== sessionId || descriptor.pauseEpoch !== pauseEpoch) return undefined;
    return { ...descriptor };
  }

  invalidateSession(sessionId: string): void {
    for (const [handle, descriptor] of this.entries) {
      if (descriptor.sessionId === sessionId) this.entries.delete(handle);
    }
  }
}

File: breakpilot-idea/src/main/kotlin/debugger/PauseScopedHandleRegistry.kt
package debugger

import com.intellij.xdebugger.frame.XValue
import java.util.UUID

data class IdeaHandleEntry(
    val sessionId: String,
    val pauseEpoch: Long,
    val value: XValue,
    val name: String,
    val frameKey: String?,
    val evaluateName: String?,
    val modifiable: Boolean
)

class PauseScopedHandleRegistry(private val maxEntries: Int = 2048) {
    private val entries = linkedMapOf<String, IdeaHandleEntry>()

    @Synchronized
    fun register(entry: IdeaHandleEntry): String {
        val handle = "bpref_${UUID.randomUUID()}"
        entries[handle] = entry
        while (entries.size > maxEntries.coerceAtLeast(1)) {
            entries.remove(entries.keys.first())
        }
        return handle
    }

    @Synchronized
    fun resolve(handle: String, sessionId: String, pauseEpoch: Long): IdeaHandleEntry? {
        val entry = entries[handle] ?: return null
        return entry.takeIf { it.sessionId == sessionId && it.pauseEpoch == pauseEpoch }
    }

    @Synchronized
    fun invalidate(sessionId: String) {
        entries.entries.removeIf { it.value.sessionId == sessionId }
    }
}
